### PR TITLE
Alglib : new package

### PR DESCRIPTION
**Pull request by davydden**
trying to figure out how to compile all `*.cpp` from python:
```
==> '/Users/davydden/spack/lib/spack/env/cc' '-c' '-fPIC' '*.cpp'
clang: error: no such file or directory: '*.cpp'
clang: error: no input files
```

### CODE REVIEW

**Comment by alalazo** [comment]
No Makefile, no CMake, no nothing? Just you, the sources and your compiler? :open_mouth:

**Comment by davydden** [comment]
> No Makefile, no CMake, no nothing? Just you, the sources and your compiler?

yes, i was shocked as well...

**Comment by alalazo** [comment]
Try to have a look at the [glob](https://docs.python.org/2.7/library/glob.html) module to get the list of files.

**Comment by adamjstewart** [comment]
Can you convert this to MakefilePackage?

**Comment by davydden** [comment]
> Can you convert this to MakefilePackage?

i could, but i still need to manually copy homemade `Makefile` as well as manually install the lib.

**Comment by adamjstewart** [comment]
Yes, you can do that during the edit phase.

File: var/spack/repos/builtin/packages/alglib/package.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        make_file = join_path(self.stage.source_path, 'src', 'Makefile')
+        shutil.copy(make_file_src, make_file)
+        filter_file(r'so', dso_suffix, make_file)
+
+    def install(self, spec, prefix):
+        name = 'libAlglib.{0}'.format(dso_suffix)
+        with working_dir('src'):
+            mkdirp(self.prefix.include)
+            headers = glob.glob(join_path(
+                self.stage.source_path, 'src', '*.h'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/alglib/package.py]: You don't need to use `self.stage.source_path` here, relative paths are fine. `*.h` should be enough.

File: var/spack/repos/builtin/packages/alglib/package.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            for h in headers:
+                install(h, self.prefix.include)
+
+            mkdirp(self.prefix.lib)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/alglib/package.py]: Don't need to use `self`, you already have access to `prefix`.

**Comment by davydden** [comment]
> I wonder if it would be easier to add an install target to the Makefile.

could be, but i don't know much about makefiles :smile:

File: var/spack/repos/builtin/packages/alglib/package.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    def edit(self, spec, prefix):
+        # this package has no build system! Nada!
+        make_file_src = join_path(os.path.dirname(self.module.__file__),
+                                  'Makefile')
+        make_file = join_path(self.stage.source_path, 'src', 'Makefile')
+        shutil.copy(make_file_src, make_file)
+        filter_file(r'so', dso_suffix, make_file)
+
+    def install(self, spec, prefix):
+        name = 'libAlglib.{0}'.format(dso_suffix)
```
Comment (davydden) [inline var/spack/repos/builtin/packages/alglib/package.py]: i am not sure if it should be `libAlglib` or `libalglib`? Trying to find out how Ubuntu call it...

Comment (davydden) [inline var/spack/repos/builtin/packages/alglib/package.py]: ok, looks like they don't capitalize https://packages.ubuntu.com/trusty/all/libalglib-dev/filelist

File: var/spack/repos/builtin/packages/alglib/package.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        shutil.copy(make_file_src, make_file)
+        filter_file(r'so', dso_suffix, make_file)
+
+    def install(self, spec, prefix):
+        name = 'libalglib.{0}'.format(dso_suffix)
+        with working_dir('src'):
+            mkdirp(self.prefix.include)
+            headers = glob.glob('*.h')
+            for h in headers:
+                install(h, self.prefix.include)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/alglib/package.py]: You can use `prefix.include` here, no need for self. Elsewhere too.

**Comment by becker33** [comment]
Seems reasonable enough as a way of dealing with their "build system".

**Pull request merged**